### 01_Set-up
Followed two Medium posts to set up a new env and install llama3.

More instructions here on downloading llama3 <br>
https://medium.com/@kagglepro.llc/easy-guide-to-installing-llama-3-by-meta-8fe5c195af39

How to add env to notebook<br>
https://medium.com/@k.yara/how-to-set-up-a-conda-environment-with-a-jupyter-kernel-f9c963207491

```
conda create -n llama3 python=3.8
conda activate llama3

pip install torch transformers

conda install -c conda-forge scikit-learn pandas

conda install -c conda-forge ipykernel

python -m ipykernel install --user --name=llama3

#to remove an env
conda env remove --name myenv

```

### Work through these warnings and issues

For notebook_login() from hugginface needed these widgets added to env
```
pip install ipywidgets
```

I installed these but got a warning about pydantic and reinstalled my previous version in this env 

```
pip install llama-index
pip install llama-index-llms-huggingface
pip install llama-index-embeddings-huggingface

```
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour isthe source of the following dependency conflicts.
text-generation 0.7.0 requires pydantic<3,>2, but you have pydantic 1.10.15 which is incompatible.

I put this back to what I had
```
pip install pydantic=2.7.1
```


#### accelerate
https://pypi.org/project/accelerate/ <br>

I'm not sure what this does but needed it to download the llm using HuggingFaceLLM

```
pip install accelerate

```

In [1]:
import torch
print(torch.version.cuda)

12.1


**Note** This installation had a cuda version incompatable with my driver, worked out below. 

#### GPU Cpmpatability
I should have GPUs but I think I need an older version of torch that works with my version of CudaToolKit

This link has what versions to use across pytorch and cudatoolkit. \
After running nvcc --version I saw I had cuda 10.1 \
Note this is not the same as shown from nvidia-smi. \
\
I installed the packages from the link based on CUDA 10.1 \
<https://pytorch.org/get-started/previous-versions/>
```
# CUDA 10.1
conda install pytorch==1.7.1 torchvision==0.8.2 torchaudio==0.7.2 cudatoolkit=10.1 -c pytorch
```

In [1]:
import torch
print(torch.__version__)

1.7.1


In [2]:
torch.cuda.is_available()

True

In [5]:
torch.cuda.device_count() 

3

In [3]:
print(torch.cuda.get_arch_list())

['sm_37', 'sm_50', 'sm_60', 'sm_61', 'sm_70', 'sm_75', 'compute_37']


In [6]:
print(torch.version.cuda)

10.1


#### Install more recent pytorch with right cuda.
pytorch 1.7.1 is compatable with CUDA 10.1, but I have driver 470.x so should be able to run a more recent version of pytorch

In [8]:
#get the nvidia GPU driver
#https://stackoverflow.com/questions/13125714/how-to-get-the-nvidia-driver-version-from-the-command-line
!cat /proc/driver/nvidia/version

/bin/bash: /home/mtseman/miniconda3/lib/libtinfo.so.6: no version information available (required by /bin/bash)
NVRM version: NVIDIA UNIX x86_64 Kernel Module  470.239.06  Sat Feb  3 06:03:07 UTC 2024
GCC version:  gcc version 9.4.0 (Ubuntu 9.4.0-1ubuntu1~20.04.2) 


In [1]:
import torch
print(torch.__version__)

2.2.2


In [2]:
print(torch.version.cuda)

11.8


#### Conclusion
This makes sense now. The link here shows my driver at 470.x works for 11.x. But initially when I installed pytorch from the links at the top, I had CUDA with torch.versio.cuda of 12.1. My driver is only good to 11.x.

https://docs.nvidia.com/deploy/cuda-compatibility/#binary-compatibility

In [3]:
torch.cuda.is_available()

True